pkg/scheduler/queue/queue.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+// Package schedulingqueue features a scheduling queue, which keeps track of all placements for the scheduler
+// to schedule.
+package schedulingqueue
+
+import (
+	"k8s.io/client-go/util/workqueue"
+)
+
+// PolicySnapshotKey is the unique identifier for a PolicySnapshot stored in a scheduling queue.
+type PolicySnapshotKey string
+
+// PolicySnapshotKeySchedulingQueueWriter is an interface which allows sources, such as controllers, to add
+// PolicySnapshots to the scheduling queue.
+type PolicySnapshotKeySchedulingQueueWriter interface {
+	Add(cpsKey PolicySnapshotKey)
+}
+
+// PolicySnapshotSchedulingQueue is an interface which queues PolicySnapshots for the scheduler to schedule.
+type PolicySnapshotKeySchedulingQueue interface {
```

Comment:
is the queue designed to handle the crp policy only?

how do we handle the resourcePlacement in the future?

Reply:
This is a scheduler only queue (in the scheduler package), and the scheduler cares only for PolicySnapshots IIRC.

Reply:
we don't need to implement it now but still need to think about it when designing the interface and at least make the queue interface easy to extend in the future.

Right now, some of the definitions are generic, eg, "PolicySnapshotKey" and some of them are specific, eg, "NextClusterPolicySnapshotKey".

Should we make them generic and later we could have two queues?

WDYT?

Reply:
I am not sure I followed. The API is literally called `ClusterPolicySnapshot`, and at this moment it is the only policy related API we have; do you mean that we will have a different API for `ResourcePlacement` (if we use that name) policy snapshots? have we made that decision yet?

I don't particularly mind changing the names, but do allow me to explain my stance a little bit clearer:

Firstly, we do not have `ResourcePlacement` yet. We do not know how it will evolve, and we definitely haven't thought it through. There might be many unexpected inconsistencies we need to address, and right now it is not even on the list. Premature accommodations can be harmful.

Secondly, this is an internal API and we can easily change it will a simple PR if the situation arrives. This is not a forward compatibility issue per se. I do believe that right now we should focus on the tasks on hand; I am happy to send another PR to add an interface or make the current one more generic if appropriate.

Reply:
`do you mean that we will have a different API for ResourcePlacement (if we use that name) policy snapshots? `

yeah, i think so, cause `ClusterPolicySnapshot` is a cluster scope resource and later we need a namespace scope resource for the "ResourcePlacement".

@ryanzhang-oss What's your preference?

Reply:
I probably haven't made myself clear enough, sorry about that, but my point is not about whether we will have a new API, but that we do NOT need to think about this in a premature fashion. We can easily make changes when the situation arises with no side effects at all.

Reply:
I think there is a possibility that we will support namespaced placement. However, the interesting thing here is that the interface is pretty much generic except one function "NextClusterPolicySnapshotKey". Everything else uses PolicySnapshot instead of ClusterPolicySnapshot. 

I think there is a chance we can unify ClusterPolicySnapshot and PolicySnapshot as both will have the exact same spec (I didn't change the spec name, just the type name). The only difference is the policyKey would have a namespace in it. With that said, I think it's also Okey to not do forward design but the current interface seems to be in between.

Reply:
In that case, would it be OK if I use the full name `ClusterPolicySnapshot` all along? We can then make changes (e.g., make things more generic if needed) when the situation matures.

Reply:
i;m ok with using full name ClusterPolicySnapshot all along and enhancing later. if the code state is in the middle, it will cause some confusions.